overlord/snapstate/conflict.go

```diff
@@ -135,6 +141,7 @@ func changeIsSnapdDowngrade(st *state.State, chg *state.Change) (bool, error) {
 	// On older snapd's 'Version' might be empty, and in this case we assume
 	// that snapd is downgrading as we cannot determine otherwise.
 	if snapsup.Version == "" {
+		fmt.Printf("@@@ This is a snapd downgrade!!!\n")
```

Comment:
This was introduced [here](https://github.com/snapcore/snapd/pull/12826/files#diff-122c976502b48247f38d9cdb27ed93c2a1186505bcdba6b5223e1be5aa02d210R137-R139).

The snapd snapsetup with version == "" comes from a previous version changing to this code which is the later version. The changes to add the version to snapsetup and to add this check was merged on 23 and 30 June 2023.